crates/utils/signposter/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+use std::{
+    ffi::{CStr, CString},
+    ptr::null,
+};
+
+use signposter_sys::*;
+
+#[cfg(feature = "global")]
+mod global;
+
+#[cfg(feature = "global")]
+pub use global::*;
+
+pub enum LogCategory {
+    /// Signposts provide high-level events that can be used to orient a
+    /// developer looking at performance data. These will be displayed by
+    /// default by performance tools like Instruments.app.
+    PointsOfInterest,
+    /// Signposts should be disabled by default, reducing the runtime overhead.
+    /// [`Log::signpost_enabled`] calls will only return 'true' when a
+    /// performance tool like Instruments.app is recording.
+    DynamicTracing,
+    /// Signposts should capture user backtraces. This behavior is more
+    /// expensive, so [`Log::signpost_enabled`] will only return 'true' when
+    /// a performance tool like Instruments.app is recording.
+    DynamicStackTracing,
+}
+
+impl LogCategory {
+    fn as_c_str(&self) -> &CStr {
+        match self {
+            LogCategory::PointsOfInterest => {
+                CStr::from_bytes_with_nul(b"PointsOfInterest\0").unwrap()
+            }
+            LogCategory::DynamicTracing => CStr::from_bytes_with_nul(b"DynamicTracing\0").unwrap(),
+            LogCategory::DynamicStackTracing => {
+                CStr::from_bytes_with_nul(b"DynamicStackTracing\0").unwrap()
+            }
+        }
+    }
+}
+
+/// A log is a collection of signposts.
+#[derive(Debug, Clone, Copy)]
+pub struct Log {
+    os_log: *mut os_log_s,
+}
+
+// Safety: unclear
+unsafe impl Sync for Log {}
+// Safety: unclear
+unsafe impl Send for Log {}
+
+impl Default for Log {
+    fn default() -> Self {
+        Log {
+            os_log: unsafe { &_os_log_default as *const _ as *mut _ },
+        }
+    }
+}
+
+impl Log {
+    /// Create a new log with the given subsystem and category.
+    pub fn new_with_subsystem_and_category(subsystem: &str, category: LogCategory) -> Self {
+        let subsystem = CString::new(subsystem).unwrap();
```

Comment:
Same here, the content of `subsystem` is not dropped while the function returns

Reply:
Can you explain what you mean here?